File: rpc/src/v1/traits/eth.rs
// Copyright 2015-2018 Parity Technologies (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

//! Eth rpc interface.
use jsonrpc_core::{Result, BoxFuture};
use jsonrpc_macros::Trailing;

use v1::types::{RichBlock, BlockNumber, Bytes, CallRequest, Filter, FilterChanges, Index};
use v1::types::{Log, Receipt, SyncStatus, Transaction, Work, SubmitDetailResult};
use v1::types::{H64, H160, H256, U256};

build_rpc_trait! {
	/// Eth rpc interface.
	pub trait Eth {
		type Metadata;

		/// Returns protocol version encoded as a string (quotes are necessary).
		#[rpc(name = "eth_protocolVersion")]
		fn protocol_version(&self) -> Result<String>;

		/// Returns an object with data about the sync status or false. (wtf?)
		#[rpc(name = "eth_syncing")]
		fn syncing(&self) -> Result<SyncStatus>;

		/// Returns the number of hashes per second that the node is mining with.
		#[rpc(name = "eth_hashrate")]
		fn hashrate(&self) -> Result<U256>;

		/// Returns block author.
		#[rpc(name = "eth_coinbase")]
		fn author(&self) -> Result<H160>;

		/// Returns true if client is actively mining new blocks.
		#[rpc(name = "eth_mining")]
		fn is_mining(&self) -> Result<bool>;

		/// Returns current gas_price.
		#[rpc(name = "eth_gasPrice")]
		fn gas_price(&self) -> Result<U256>;

		/// Returns accounts list.
		#[rpc(name = "eth_accounts")]
		fn accounts(&self) -> Result<Vec<H160>>;

		/// Returns highest block number.
		#[rpc(name = "eth_blockNumber")]
		fn block_number(&self) -> Result<U256>;

		/// Returns balance of the given account.
		#[rpc(name = "eth_getBalance")]
		fn balance(&self, H160, Trailing<BlockNumber>) -> BoxFuture<U256>;

		/// Returns content of the storage at given address.
		#[rpc(name = "eth_getStorageAt")]
		fn storage_at(&self, H160, U256, Trailing<BlockNumber>) -> BoxFuture<H256>;

		/// Returns block with given hash.
		#[rpc(name = "eth_getBlockByHash")]
		fn block_by_hash(&self, H256, bool) -> BoxFuture<Option<RichBlock>>;

		/// Returns block with given number.
		#[rpc(name = "eth_getBlockByNumber")]
		fn block_by_number(&self, BlockNumber, bool) -> BoxFuture<Option<RichBlock>>;

		/// Returns the number of transactions sent from given address at given time (block number).
		#[rpc(name = "eth_getTransactionCount")]
		fn transaction_count(&self, H160, Trailing<BlockNumber>) -> BoxFuture<U256>;

		/// Returns the number of transactions in a block with given hash.
		#[rpc(name = "eth_getBlockTransactionCountByHash")]
		fn block_transaction_count_by_hash(&self, H256) -> BoxFuture<Option<U256>>;

		/// Returns the number of transactions in a block with given block number.
		#[rpc(name = "eth_getBlockTransactionCountByNumber")]
		fn block_transaction_count_by_number(&self, BlockNumber) -> BoxFuture<Option<U256>>;

		/// Returns the number of uncles in a block with given hash.
		#[rpc(name = "eth_getUncleCountByBlockHash")]
		fn block_uncles_count_by_hash(&self, H256) -> BoxFuture<Option<U256>>;

		/// Returns the number of uncles in a block with given block number.
		#[rpc(name = "eth_getUncleCountByBlockNumber")]
		fn block_uncles_count_by_number(&self, BlockNumber) -> BoxFuture<Option<U256>>;

		/// Returns the code at given address at given time (block number).
		#[rpc(name = "eth_getCode")]
		fn code_at(&self, H160, Trailing<BlockNumber>) -> BoxFuture<Bytes>;

		/// Sends signed transaction, returning its hash.
		#[rpc(name = "eth_sendRawTransaction")]
		fn send_raw_transaction(&self, Bytes) -> Result<H256>;

		/// @alias of `eth_sendRawTransaction`.
		#[rpc(name = "eth_submitTransaction")]
		fn submit_transaction(&self, Bytes) -> Result<H256>;

		/// Call contract, returning the output data.
		#[rpc(name = "eth_call")]
		fn call(&self, CallRequest, Trailing<BlockNumber>) -> BoxFuture<Bytes>;

		/// Estimate gas needed for execution of given contract.
		#[rpc(name = "eth_estimateGas")]
		fn estimate_gas(&self, CallRequest, Trailing<BlockNumber>) -> BoxFuture<U256>;

		/// Get transaction by its hash.
		#[rpc(name = "eth_getTransactionByHash")]
		fn transaction_by_hash(&self, H256) -> BoxFuture<Option<Transaction>>;

		/// Returns transaction at given block hash and index.
		#[rpc(name = "eth_getTransactionByBlockHashAndIndex")]
		fn transaction_by_block_hash_and_index(&self, H256, Index) -> BoxFuture<Option<Transaction>>;

		/// Returns transaction by given block number and index.
		#[rpc(name = "eth_getTransactionByBlockNumberAndIndex")]
		fn transaction_by_block_number_and_index(&self, BlockNumber, Index) -> BoxFuture<Option<Transaction>>;

		/// Returns transaction receipt by transaction hash.
		#[rpc(name = "eth_getTransactionReceipt")]
		fn transaction_receipt(&self, H256) -> BoxFuture<Option<Receipt>>;

		/// Returns an uncles at given block and index.
		#[rpc(name = "eth_getUncleByBlockHashAndIndex")]
		fn uncle_by_block_hash_and_index(&self, H256, Index) -> BoxFuture<Option<RichBlock>>;

		/// Returns an uncles at given block and index.
		#[rpc(name = "eth_getUncleByBlockNumberAndIndex")]
		fn uncle_by_block_number_and_index(&self, BlockNumber, Index) -> BoxFuture<Option<RichBlock>>;

		/// Returns available compilers.
		/// @deprecated
		#[rpc(name = "eth_getCompilers")]
		fn compilers(&self) -> Result<Vec<String>>;

		/// Compiles lll code.
		/// @deprecated
		#[rpc(name = "eth_compileLLL")]
		fn compile_lll(&self, String) -> Result<Bytes>;

		/// Compiles solidity.
		/// @deprecated
		#[rpc(name = "eth_compileSolidity")]
		fn compile_solidity(&self, String) -> Result<Bytes>;

		/// Compiles serpent.
		/// @deprecated
		#[rpc(name = "eth_compileSerpent")]
		fn compile_serpent(&self, String) -> Result<Bytes>;

		/// Returns logs matching given filter object.
		#[rpc(name = "eth_getLogs")]
		fn logs(&self, Filter) -> BoxFuture<Vec<Log>>;

		/// Returns the hash of the current block, the seedHash, and the boundary condition to be met.
		#[rpc(name = "eth_getWork")]
		fn work(&self, Trailing<u64>) -> Result<Work>;

		/// Used for submitting a proof-of-work solution.
		#[rpc(name = "eth_submitWork")]
		fn submit_work(&self, H64, H256, H256) -> Result<bool>;

		/// Used for submitting a proof-of-work solution (more details in the response than `eth_submitWork`).
		#[rpc(name = "eth_submitWorkDetail")]
		fn submit_work_detail(&self, H64, H256, H256) -> Result<SubmitDetailResult>;

		/// Used for submitting mining hashrate.
		#[rpc(name = "eth_submitHashrate")]
		fn submit_hashrate(&self, U256, H256) -> Result<bool>;
	}
}

build_rpc_trait! {
	/// Eth filters rpc api (polling).
	// TODO: do filters api properly
	pub trait EthFilter {
		/// Returns id of new filter.
		#[rpc(name = "eth_newFilter")]
		fn new_filter(&self, Filter) -> Result<U256>;

		/// Returns id of new block filter.
		#[rpc(name = "eth_newBlockFilter")]
		fn new_block_filter(&self) -> Result<U256>;

		/// Returns id of new block filter.
		#[rpc(name = "eth_newPendingTransactionFilter")]
		fn new_pending_transaction_filter(&self) -> Result<U256>;

		/// Returns filter changes since last poll.
		#[rpc(name = "eth_getFilterChanges")]
		fn filter_changes(&self, Index) -> BoxFuture<FilterChanges>;

		/// Returns all logs matching given filter (in a range 'from' - 'to').
		#[rpc(name = "eth_getFilterLogs")]
		fn filter_logs(&self, Index) -> BoxFuture<Vec<Log>>;

		/// Uninstalls filter.
		#[rpc(name = "eth_uninstallFilter")]
		fn uninstall_filter(&self, Index) -> Result<bool>;
	}
}

File: js-old/src/api/pubsub/eth/eth.js
// Copyright 2015-2017 Parity Technologies (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
import PubsubBase from '../pubsubBase';

import { inAddress, inBlockNumber, inHex, inNumber16, inOptions, inFilter } from '../../format/input';
import { outAddress, outBlock, outNumber, outTransaction, outSyncing, outReceipt, outLog } from '../../format/output';

export default class Eth extends PubsubBase {
  constructor (transport) {
    super(transport);
    this._api = 'parity';
  }

  newHeads (callback) {
    return this.addListener('eth', 'newHeads', callback);
  }

  logs (callback) {
    throw Error('not supported yet');
  }

  //  eth API
  protocolVersion (callback) {
    return this.addListener(this._api, 'eth_protocolVersion', callback);
  }

  syncing (callback) {
    return this.addListener(this._api, 'eth_syncing', (error, data) => {
      error
        ? callback(error)
        : callback(null, outSyncing(data));
    });
  }

  hashrate (callback) {
    return this.addListener(this._api, 'eth_hashrate', (error, data) => {
      error
        ? callback(error)
        : callback(null, outNumber(data));
    });
  }

  coinbase (callback) {
    return this.addListener(this._api, 'eth_coinbase', (error, data) => {
      error
        ? callback(error)
        : callback(null, outAddress(data));
    });
  }

  mining (callback) {
    return this.addListener(this._api, 'eth_mining', callback);
  }

  gasPrice (callback) {
    return this.addListener(this._api, 'eth_gasPrice', (error, data) => {
      error
        ? callback(error)
        : callback(null, outNumber(data));
    });
  }

  accounts (callback) {
    return this.addListener(this._api, 'eth_accounts', (error, accounts) => {
      error
        ? callback(error)
        : callback(null, (accounts || []).map(outAddress));
    });
  }

  blockNumber (callback) {
    return this.addListener(this._api, 'eth_blockNumber', (error, data) => {
      error
        ? callback(error)
        : callback(null, outNumber(data));
    });
  }

  getBalance (callback, address, blockNumber = 'latest') {
    return this.addListener(this._api, 'eth_getBalance', (error, data) => {
      error
        ? callback(error)
        : callback(null, outNumber(data));
    }, [inAddress(address), inBlockNumber(blockNumber)]);
  }

  getStorageAt (callback, address, index = 0, blockNumber = 'latest') {
    return this.addListener(this._api, 'eth_getStorageAt', callback, [inAddress(address), inNumber16(index), inBlockNumber(blockNumber)]);
  }

  getBlockByHash (callback, hash, full = false) {
    return this.addListener(this._api, 'eth_getBlockByHash', (error, data) => {
      error
        ? callback(error)
        : callback(null, outBlock(data));
    }, [inHex(hash), full]);
  }

  getBlockByNumber (callback, blockNumber = 'latest', full = false) {
    return this.addListener(this._api, 'eth_getBlockByNumber', (error, data) => {
      error
        ? callback(error)
        : callback(null, outBlock(data));
    }, [inBlockNumber(blockNumber), full]);
  }

  getTransactionCount (callback, address, blockNumber = 'latest') {
    return this.addListener(this._api, 'eth_getTransactionCount', (error, data) => {
      error
        ? callback(error)
        : callback(null, outNumber(data));
    }, [inAddress(address), inBlockNumber(blockNumber)]);
  }

  getBlockTransactionCountByHash (callback, hash) {
    return this.addListener(this._api, 'eth_getBlockTransactionCountByHash', (error, data) => {
      error
        ? callback(error)
        : callback(null, outNumber(data));
    }, [inHex(hash)]);
  }

  getBlockTransactionCountByNumber (callback, blockNumber = 'latest') {
    return this.addListener(this._api, 'eth_getBlockTransactionCountByNumber', (error, data) => {
      error
        ? callback(error)
        : callback(null, outNumber(data));
    }, [inBlockNumber(blockNumber)]);
  }

  getUncleCountByBlockHash (callback, hash) {
    return this.addListener(this._api, 'eth_getUncleCountByBlockHash', (error, data) => {
      error
        ? callback(error)
        : callback(null, outNumber(data));
    }, [inHex(hash)]);
  }

  getUncleCountByBlockNumber (callback, blockNumber = 'latest') {
    return this.addListener(this._api, 'eth_getUncleCountByBlockNumber', (error, data) => {
      error
        ? callback(error)
        : callback(null, outNumber(data));
    }, [inBlockNumber(blockNumber)]);
  }

  getCode (callback, address, blockNumber = 'latest') {
    return this.addListener(this._api, 'eth_getCode', callback, [inAddress(address), inBlockNumber(blockNumber)]);
  }

  call (callback, options, blockNumber = 'latest') {
    return this.addListener(this._api, 'eth_call', callback, [inOptions(options), inBlockNumber(blockNumber)]);
  }

  estimateGas (callback, options) {
    return this.addListener(this._api, 'eth_estimateGas', (error, data) => {
      error
        ? callback(error)
        : callback(null, outNumber(data));
    }, [inOptions(options)]);
  }

  getTransactionByHash (callback, hash) {
    return this.addListener(this._api, 'eth_getTransactionByHash', (error, data) => {
      error
        ? callback(error)
        : callback(null, outTransaction(data));
    }, [inHex(hash)]);
  }

  getTransactionByBlockHashAndIndex (callback, hash, index = 0) {
    return this.addListener(this._api, 'eth_getTransactionByBlockHashAndIndex', (error, data) => {
      error
        ? callback(error)
        : callback(null, outTransaction(data));
    }, [inHex(hash), inNumber16(index)]);
  }

  getTransactionByBlockNumberAndIndex (callback, blockNumber = 'latest', index = 0) {
    return this.addListener(this._api, 'eth_getTransactionByBlockNumberAndIndex', (error, data) => {
      error
        ? callback(error)
        : callback(null, outTransaction(data));
    }, [inBlockNumber(blockNumber), inNumber16(index)]);
  }

  getTransactionReceipt (callback, txhash) {
    return this.addListener(this._api, 'eth_getTransactionReceipt', (error, data) => {
      error
        ? callback(error)
        : callback(null, outReceipt(data));
    }, [inHex(txhash)]);
  }

  getUncleByBlockHashAndIndex (callback, hash, index = 0) {
    return this.addListener(this._api, 'eth_getUncleByBlockHashAndIndex', callback, [inHex(hash), inNumber16(index)]);
  }

  getUncleByBlockNumberAndIndex (callback, blockNumber = 'latest', index = 0) {
    return this.addListener(this._api, 'eth_getUncleByBlockNumberAndIndex', callback, [inBlockNumber(blockNumber), inNumber16(index)]);
  }

  getLogs (callback, options) {
    return this.addListener(this._api, 'eth_getLogs', (error, logs) => {
      error
        ? callback(error)
        : callback(null, (logs) => logs.map(outLog));
    }, [inFilter(options)]);
  }

  getWork (callback) {
    return this.addListener(this._api, 'eth_getWork', callback);
  }
}
